.github/workflows/ci.yml

```diff
@@ -16,23 +16,35 @@ env:
 jobs:
   test:
     name: Test ${{ matrix.crate }}
-    runs-on: ubuntu-latest
+    runs-on: ${{ fromJSON(
+      github.repository == 'libp2p/rust-libp2p' && (
+        (contains(fromJSON('["libp2p-webrtc", "libp2p"]'), matrix.crate) && '["self-hosted", "linux", "x64", "2xlarge"]') ||
+        (contains(fromJSON('["libp2p-quic", "libp2p-perf"]'), matrix.crate) && '["self-hosted", "linux", "x64", "xlarge"]') ||
+        '["self-hosted", "linux", "x64", "large"]'
+      ) || '"ubuntu-latest"') }}
+    timeout-minutes: 10
     needs: gather_published_crates
     strategy:
       fail-fast: false
       matrix:
         crate: ${{ fromJSON(needs.gather_published_crates.outputs.members) }}
     steps:
       - name: Install Protoc
-        run: sudo apt-get install protobuf-compiler
+        run: sudo apt-get install -y protobuf-compiler
+
+      - if: github.repository == 'libp2p/rust-libp2p'
+        uses: dtolnay/rust-toolchain@stable
 
       - uses: actions/checkout@v3
 
       - name: Get MSRV for ${{ matrix.crate }}
         id: parse-msrv
+        env:
+          CRATE: ${{ matrix.crate }}
```

Comment:
Generally, I prefer not to because to me specifying env per step gives me explicit visibility into some of the step's dependencies. It's just a matter of preference though so feel free to change it.

Not using template expansion - `${{}}` - inside run blocks is considered safer because you can do code injection through templates. Here, with the current set of triggers and all, it doesn't matter that much because the code injected through a crate name could just be more easily introduced through another run block. But I usually make this change whenever I go through a new workflow yml just not to even have to think about it later.

I'm travelling today so won't be able to get to it myself.

Reply:
Thanks, I'll send a separate PR!

Reply:
FYI, I removed the env changes from this one.